x-pack/filebeat/input/azureeventhub/input.go

```diff
@@ -0,0 +1,223 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package azureeventhub
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/filebeat/channel"
+	"github.com/elastic/beats/filebeat/input"
+	"github.com/elastic/beats/libbeat/beat"
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/logp"
+
+	eventhub "github.com/Azure/azure-event-hubs-go/v3"
+	"github.com/Azure/azure-event-hubs-go/v3/eph"
+)
+
+const (
+	eventHubConnector        = ";EntityPath="
+	expandEventListFromField = "records"
+)
+
+// azureInput struct for the azure-eventhub input
+type azureInput struct {
+	config       azureInputConfig // azure-eventhub configuration
+	context      input.Context
+	outlet       channel.Outleter
+	log          *logp.Logger            // logging info and error messages
+	workerCtx    context.Context         // worker goroutine context. It's cancelled when the input stops or the worker exits.
+	workerCancel context.CancelFunc      // used to signal that the worker should stop.
+	workerOnce   sync.Once               // guarantees that the worker goroutine is only started once.
+	workerWg     sync.WaitGroup          // waits on worker goroutine.
+	processor    *eph.EventProcessorHost // eph will be assigned if users have enabled the option
+	hub          *eventhub.Hub           // hub will be assigned
+}
+
+const (
+	inputName = "azure-eventhub"
+)
+
+func init() {
+	err := input.Register(inputName, NewInput)
+	if err != nil {
+		panic(errors.Wrapf(err, "failed to register %v input", inputName))
+	}
+}
+
+// NewInput creates a new azure-eventhub input
+func NewInput(
+	cfg *common.Config,
+	connector channel.Connector,
+	inputContext input.Context,
+) (input.Input, error) {
+	var config azureInputConfig
+	if err := cfg.Unpack(&config); err != nil {
+		return nil, errors.Wrapf(err, "reading %s input config", inputName)
+	}
+	out, err := connector.ConnectWith(cfg, beat.ClientConfig{
+		Processing: beat.ProcessingConfig{
+			DynamicFields: inputContext.DynamicFields,
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	inputCtx, cancelInputCtx := context.WithCancel(context.Background())
+	go func() {
+		defer cancelInputCtx()
+		select {
+		case <-inputContext.Done:
+		case <-inputCtx.Done():
```

Comment:
this one should not be needed right? As you use this to cancel `inputCtx`

Reply:
nevermind, you need it to avoid leaking this one

Reply:
yes